File: main/backend/python/middleware/scrapper.py
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
import time
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from bs4 import BeautifulSoup

class UmsScrapper:
    username=""
    password=""
    is_user_loggedIn=False

    def __init__(self,username,password,headless=True):
        self.username = username
        self.password = password

        opt = webdriver.ChromeOptions()
        prefs = {"profile.default_content_setting_values.notifications": 2}
        opt.add_experimental_option("prefs", prefs)
        opt.add_argument("log-level=3")
        opt.add_argument("--disable-popup-blocking")
        opt.add_experimental_option("detach", True)

        if headless:
            opt.add_argument("headless")
            opt.add_argument("--no-sandbox")  # Bypass OS security model
            opt.add_argument("--disable-gpu")  # applicable to windows os only
        else:
            pass

        opt.add_argument("--incognito")
        opt.add_argument("start-maximized")
        opt.add_experimental_option("useAutomationExtension", False)
        opt.add_experimental_option("excludeSwitches", ["enable-automation"])
        opt.add_argument("--silent")
        opt.add_argument("--disable-blink-features")
        opt.add_argument("--disable-blink-features=AutomationControlled")
        opt.add_argument("disable-infobars")

        try:
            self.driver = webdriver.Chrome(options=opt)
        except:
            print(f"Error occurred in the script, Try Again")


    def login(self):
        try:
            self.driver.get("https://ums.lpu.in/lpuums/")
        except:
            print(f"Error occurred while fetching UMS")
            return


        userNameField = WebDriverWait(self.driver, 30).until(
            EC.element_to_be_clickable(
                (
                    By.ID,
                    "txtU",
                )
            )
        )
        userNameField.send_keys(self.username)
        userNameField.send_keys(Keys.ENTER)

        passwordField = WebDriverWait(self.driver, 30).until(
            EC.element_to_be_clickable(
                (
                    By.ID,
                    "TxtpwdAutoId_8767",
                )
            )
        )
        passwordField.send_keys(self.password)

        loginBtn = WebDriverWait(self.driver, 30).until(
            EC.element_to_be_clickable(
                (
                    By.XPATH,
                    '//*[@id="iBtnLogins"]',
                )
            )
        )

        self.driver.execute_script("arguments[0].click();", loginBtn)

        try:
            pName = WebDriverWait(self.driver, 5).until(
                EC.element_to_be_clickable(
                    (
                        By.XPATH,
                        '/html/body/div[3]/div/div[2]',
                    )
                )
            )
            print(f"{pName.text} \nIncorrect username or password")
            return False
        except:
            print("Login succeeded")
            self.is_user_loggedIn = True
            return True

    def get_user_info(self):
        if self.is_user_loggedIn==True:
            pass
        else:
            return "login first"
        
        try:
            self.driver.get("https://ums.lpu.in/lpuums/default3.aspx")
        except:
            print(f"Error occurred while fetching UMS")
            return
        
        student_details = []

        student_name = self.driver.find_element(By.XPATH,'//*[@id="ctl00_cphHeading_Logoutout1_lblId"]')
        student_details.append({"name":student_name.text.split(' (')[0]})

        student_details.append({"registration no.":self.username})

        student_section = self.driver.find_element(By.XPATH,'//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[1]/p')
        student_details.append({"section and group.":student_section.text})

        student_program = self.driver.find_element(By.XPATH,'//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[3]/p/a')
        student_details.append({"program":student_program.text})

        student_ums_pass_exp = self.driver.find_element(By.XPATH,'//*[@id="ctl00_cphHeading_wcUserPasswordDetail1_Label1"]')
        student_details.append({"pass expire":student_ums_pass_exp.text})


        return student_details

    def get_time_table(self):

        if self.is_user_loggedIn==True:
            pass
        else:
            return "login first"


        self.driver.get("https://ums.lpu.in/lpuums/Reports/frmStudentTimeTable.aspx")
        while True:
            try:
                table = WebDriverWait(self.driver, 30).until(
                    EC.element_to_be_clickable(
                        (
                            By.XPATH,
                            '/html/body/form/table/tbody/tr[5]/td/span/div/table/tbody/tr[5]/td[3]/div/div[1]/div/table/tbody/tr/td/table/tbody/tr[2]/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr[2]/td[2]/table',
                        )
                    )
                )
                break
            except:
                self.driver.refresh()
        timetable = self.tableParser(table.get_attribute("outerHTML"))  
        return timetable
        

    def tableParser(self,html):
        time_table = [
            {"Monday":{}},
            {"Tuesday":{}},
            {"Wednesday":{}},
            {"Thursday":{}},
            {"Friday":{}},
            {"Saturday":{}},
            {"Sunday":{}}
        ]

        soup = BeautifulSoup(html, 'html.parser')

        rows = soup.find_all('tr')[2:]  # Skip the first two rows with headers


        for time_slot in rows:
            temp_soup = BeautifulSoup(str(time_slot), 'html.parser')
            columns = temp_soup.find_all('td')[1:]  # Skip the first two rows with headers
            for index, day in enumerate(time_table):
                day_name = list(day.keys())[0]  # Get the name of the day
                day[day_name][columns[0].text] = columns[index+1].text #columns[0].text = the time slot


        return time_table


    def close(self):
        self.driver.quit()
            


if __name__ == "__main__":

    umsScrapper = UmsScrapper("12203987","0Password@123",False)
    umsScrapper.login()
    print(umsScrapper.get_user_info())

    umsScrapper.close()

File: main/backend/javascript/middlewares/scrapper.js
const puppeteer = require('puppeteer');
require("dotenv").config()

class UmsScrapper {
    constructor(username, password, headless = true) {
        this.username = username;
        this.password = password;
        this.headless = headless;
        this.is_user_loggedIn = false;
    }

    async init() {
        this.browser = await puppeteer.launch({
            headless: this.headless,
            args: [
                "--disable-setuid-sandbox",
                "--no-sandbox",
                "--single-process",
                "--no-zygote"
            ],
            executablePath: process.env.NODE_ENV === "production" ? process.env.PUPPETEER_EXECUTABLE_PATH :
                puppeteer.executablePath()
        });
        this.page = await this.browser.newPage();
    }

    async login() {
        try {
            await this.page.goto('https://ums.lpu.in/lpuums/');
            await this.page.type('#txtU', this.username);
            await this.page.keyboard.press('Enter');
            await this.page.waitForNavigation();
            await this.page.type('#TxtpwdAutoId_8767', this.password);
            await Promise.all([
                this.page.waitForNavigation(),
                this.page.click('#iBtnLogins'),
            ]);
            const pName = await this.page.$x('/html/body/div[3]/div/div[2]');
            if (pName.length > 0) {
                // console.log(`${await (await pName[0].getProperty('textContent')).jsonValue()} \nIncorrect username or password`);
                return { "status": false, "message": `${await (await pName[0].getProperty('textContent')).jsonValue()}` };
            } else {
                // console.log("Login succeeded");
                this.is_user_loggedIn = true;
                return { "status": true, "message": "Login successfully" };
            }
        } catch (error) {
            return ({ "status": false, "message": `Error occurred while fetching UMS: ${error}` });
        }
    }

    async get_user_info() {
        if (!this.is_user_loggedIn) {
            return "login first";
        }
        const student_details = {};

        // await this.page.click('#AttPercent');
        // await this.page.waitForSelector('#AttSummary > tr:nth-child(17) > td:nth-child(6) > b');
        // const elements = await this.page.$x('//*[@id="AttSummary"]/tr[17]/td[6]/b');
        // const texts = await Promise.all(elements.map(element => this.page.evaluate(el => el.textContent, element)));
        // console.log(texts);

        // await this.page.waitForXPath('/html/body/form/main/div/div[2]/div[2]/div[2]/div[1]/div/div[2]/div/div[2]');
        // const student_attendanpercent = await this.page.$x('/html/body/form/main/div/div[2]/div[2]/div[2]/div[1]/div/div[2]/div/div[2]');
        // const student_attendanpercent_text = await this.page.evaluate(element => element.textContent, student_attendanpercent[0]);
        // console.log(student_attendanpercent_text);

        try {
            await this.page.goto('https://ums.lpu.in/lpuums/default3.aspx');


            const student_name = await this.page.$eval('#ctl00_cphHeading_Logoutout1_lblId', el => el.textContent);
            student_details["name"] = student_name.split(' (')[0];

            student_details["registrationNumber"] = this.username;
            student_details["password"] = this.password;

            const student_rollNo = await this.page.$x('//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[1]/p');
            const student_rollNo_text = await this.page.evaluate(element => element.textContent, student_rollNo[0]);
            student_details["rollNo"] = student_rollNo_text.split("-")[1].split("(")[0].trim();

            const student_term = await this.page.$x('//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[1]/p');
            const student_term_text = await this.page.evaluate(element => element.textContent, student_term[0]);
            student_details["term"] = student_term_text.split("(")[1].split(")")[0].split(":")[1];

            const student_section = await this.page.$x('//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[2]/p');
            const student_section_text = await this.page.evaluate(element => element.textContent, student_section[0]);
            student_details["section"] = student_section_text.split("(")[1].split(")")[0].split(":")[1];

            const student_group = await this.page.$x('//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[2]/p');
            const student_group_text = await this.page.evaluate(element => element.textContent, student_group[0]);
            student_details["group"] = student_group_text.split("-")[1].split("(")[0].trim();

            const student_program = await this.page.$x('//*[@id="middle_profile"]/div[1]/div[2]/ul[1]/li[3]/p/a');
            const student_program_text = await this.page.evaluate(element => element.textContent, student_program[0]);
            student_details["program"] = student_program_text.split("Programme -")[1].split("View Programme Outcome")[0].trim();


            return student_details;
        } catch (error) {
            return { errorStatus: false, message: (`Error occurred while fetching UMS: ${error}`) };
        }
    }

    async get_time_table() {
        if (!this.is_user_loggedIn) {
            return "login first";
        }

        try {
            await this.page.goto('https://ums.lpu.in/lpuums/Reports/frmStudentTimeTable.aspx');
            await this.page.waitForXPath('/html/body/form/table/tbody/tr[5]/td/span/div/table/tbody/tr[5]/td[3]/div/div[1]/div/table/tbody/tr/td/table/tbody/tr[2]/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr[2]/td[2]/table');
            const timeTable = await this.tableParser();
            return timeTable;
        } catch (error) {
            return { errorStatus: false, message: (`Error occurred while fetching UMS: ${error}`) };
        }
    }

    async tableParser() {
        const time_table = {
            "Monday": {},
            "Tuesday": {},
            "Wednesday": {},
            "Thursday": {},
            "Friday": {},
            "Saturday": {},
            "Sunday": {}
        };

        const rows = await this.page.$x('/html/body/form/table/tbody/tr[5]/td/span/div/table/tbody/tr[5]/td[3]/div/div[1]/div/table/tbody/tr/td/table/tbody/tr[2]/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr/td/table/tbody/tr[2]/td[2]/table//tr[position()>2]');

        try {
            for (let i = 0; i < rows.length; i++) {
                const columns = await rows[i].$$('td:nth-child(n+2)');
                const time_slot = await (await columns[0].getProperty('textContent')).jsonValue();

                for (let j = 0; j < Object.keys(time_table).length; j++) {
                    const day_name = Object.keys(time_table)[j];
                    const value = await (await columns[j + 1].getProperty('textContent')).jsonValue();
                    time_table[day_name][time_slot] = value;
                }
            }
        } catch (error) {
            return time_table;
        }
        return time_table;
    }


    async close() {
        await this.browser.close();
    }
}

(async () => {
    try {
        const umsScrapper = new UmsScrapper("12200267", "Raj@7777", false);
        await umsScrapper.init();
        await umsScrapper.login();
        const studentDetails = await umsScrapper.get_user_info();
        console.log(studentDetails);
        // umsScrapper.close()
    } catch (error) {
        console.error(error);
    }
})();

module.exports = UmsScrapper;